# Chapter 4: PyTorch Fundamentals

In [1]:
import torch

device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using device: {device}")

# model.to(device)

Using device: cuda


### Basic tensor construction

In [2]:
scalar = torch.tensor(42, dtype=torch.float32)
vector = torch.tensor([1, 2, 3, 4, 5], dtype=torch.float32)
matrix = torch.tensor([
 [1, 2, 3],
 [4, 5, 6],
 [7, 8, 9]
], dtype=torch.float32)

tensor_3d = torch.tensor([
 [[1, 2], [3, 4]],
 [[5, 6], [7, 8]],
 [[9, 10], [11, 12]]
], dtype=torch.float32)

print(scalar)
print(vector)
print(matrix)
print(tensor_3d)

tensor(42.)
tensor([1., 2., 3., 4., 5.])
tensor([[1., 2., 3.],
        [4., 5., 6.],
        [7., 8., 9.]])
tensor([[[ 1.,  2.],
         [ 3.,  4.]],

        [[ 5.,  6.],
         [ 7.,  8.]],

        [[ 9., 10.],
         [11., 12.]]])


Of course usually we want to create tensors with a few more numbers than we are willing to type ourselves, for specific purposes like initializing model weights, or creating placeholders etc. In which case
torch comes with some handy factory functions, such as torch.rand, torch.randn or torch.zeros:

In [19]:
torch.manual_seed(42)
randn_tensor = torch.randn(3,2)

print(f'data type \n', type(randn_tensor))
print(f'within tensor data type \n',randn_tensor.dtype)
print(f'tensor values \n', randn_tensor)

data type 
 <class 'torch.Tensor'>
within tensor data type 
 torch.float32
tensor values 
 tensor([[ 0.3367,  0.1288],
        [ 0.2345,  0.2303],
        [-1.1229, -0.1863]])


### Matrix Operations:
We’ll look at three common operations: tensor x
tensor (element-wise multiplication, where corresponding elements are multiplied together), matrix
multiplication (the dot product operation), and addition. Element-wise multiplication produces a
tensor of the same shape with each position multiplied independently, while matrix multiplication
follows linear algebra rules and often changes dimensions.
Element-wise multiplication produces a tensor of the same shape with each position multiplied independently, while matrix multiplication follows linear algebra rules and often changes dimensions.
Element-wise multiplications can be performed using either .mul() or the * operator, and matrix multiplication using either .matmul() or the @ operator. We’ll be using the method syntax here for clarity

In [20]:
print(f'tensor x tensor \n', randn_tensor.mul(randn_tensor), f'\n')
print(f'tensor x tensor \n', randn_tensor * randn_tensor, f'\n')

print(f'matrix multiplication \n', randn_tensor.matmul(randn_tensor.T))
print(f'matrix multiplication \n', randn_tensor @ randn_tensor.T)

print(f'add 1 \n', randn_tensor.add(1))
print(f'add 1 \n', randn_tensor + 1)

tensor x tensor 
 tensor([[0.1134, 0.0166],
        [0.0550, 0.0531],
        [1.2608, 0.0347]]) 

tensor x tensor 
 tensor([[0.1134, 0.0166],
        [0.0550, 0.0531],
        [1.2608, 0.0347]]) 

matrix multiplication 
 tensor([[ 0.1300,  0.1086, -0.4021],
        [ 0.1086,  0.1080, -0.3062],
        [-0.4021, -0.3062,  1.2955]])
matrix multiplication 
 tensor([[ 0.1300,  0.1086, -0.4021],
        [ 0.1086,  0.1080, -0.3062],
        [-0.4021, -0.3062,  1.2955]])
add 1 
 tensor([[ 1.3367,  1.1288],
        [ 1.2345,  1.2303],
        [-0.1229,  0.8137]])
add 1 
 tensor([[ 1.3367,  1.1288],
        [ 1.2345,  1.2303],
        [-0.1229,  0.8137]])


Some other useful operations are concatenation and simply clearing the numbers of a tensor to zero:

In [5]:
zero_tensor = torch.randn(3,2).zero_()
concat_tensor = torch.cat([zero_tensor, randn_tensor], dim = 1)
print(f'zero tensor \n', zero_tensor)
print(f'concatonated tensors \n', concat_tensor)

zero tensor 
 tensor([[0., 0.],
        [0., 0.],
        [0., 0.]])
concatonated tensors 
 tensor([[ 0.0000,  0.0000,  1.3367,  1.1288],
        [ 0.0000,  0.0000,  1.2345,  1.2303],
        [ 0.0000,  0.0000, -0.1229,  0.8137]])


At this point we know how to construct a tensor, but you’ll also need to convert data from DataFrame.
The simplest approach is to pull the values of the columns you’re interested in. If you have a single
column, this will just create a 1-dimensional tensor, while additional columns will create a 2-dimensional tensor.

Let’s look at a practical example using the airline passengers dataset. We’ll demonstrate how to convert
pandas DataFrame columns into PyTorch tensors, creating both a univariate tensor (single column)
and a multivariate tensor (multiple columns).

In [10]:
import numpy as np
import pandas as pd
from Utilities.PathUtilities import get_current_working_directory

# Load dataset
path = get_current_working_directory().parent / 'Data' / 'passengers.csv'
data = pd.read_csv(path)

# Add a count column
data['count'] = np.arange(1, len(data) + 1)
tensor_1_col = torch.tensor(data['passengers'].values,
 dtype=torch.float32)
tensor_2_col = torch.tensor(data[['passengers', 'count']].values,
 dtype=torch.float32)

print(f'simple univariate tensor')
tensor_1_col

simple univariate tensor


tensor([112., 118., 132., 129., 121., 135., 148., 148., 136., 119., 104., 118.,
        115., 126., 141., 135., 125., 149., 170., 170., 158., 133., 114., 140.,
        145., 150., 178., 163., 172., 178., 199., 199., 184., 162., 146., 166.,
        171., 180., 193., 181., 183., 218., 230., 242., 209., 191., 172., 194.,
        196., 196., 236., 235., 229., 243., 264., 272., 237., 211., 180., 201.,
        204., 188., 235., 227., 234., 264., 302., 293., 259., 229., 203., 229.,
        242., 233., 267., 269., 270., 315., 364., 347., 312., 274., 237., 278.,
        284., 277., 317., 313., 318., 374., 413., 405., 355., 306., 271., 306.,
        315., 301., 356., 348., 355., 422., 465., 467., 404., 347., 305., 336.,
        340., 318., 362., 348., 363., 435., 491., 505., 404., 359., 310., 337.,
        360., 342., 406., 396., 420., 472., 548., 559., 463., 407., 362., 405.,
        417., 391., 419., 461., 472., 535., 622., 606., 508., 461., 390., 432.])

In [11]:
print(f'simple multivariate tensor')
tensor_2_col

simple multivariate tensor


tensor([[112.,   1.],
        [118.,   2.],
        [132.,   3.],
        [129.,   4.],
        [121.,   5.],
        [135.,   6.],
        [148.,   7.],
        [148.,   8.],
        [136.,   9.],
        [119.,  10.],
        [104.,  11.],
        [118.,  12.],
        [115.,  13.],
        [126.,  14.],
        [141.,  15.],
        [135.,  16.],
        [125.,  17.],
        [149.,  18.],
        [170.,  19.],
        [170.,  20.],
        [158.,  21.],
        [133.,  22.],
        [114.,  23.],
        [140.,  24.],
        [145.,  25.],
        [150.,  26.],
        [178.,  27.],
        [163.,  28.],
        [172.,  29.],
        [178.,  30.],
        [199.,  31.],
        [199.,  32.],
        [184.,  33.],
        [162.,  34.],
        [146.,  35.],
        [166.,  36.],
        [171.,  37.],
        [180.,  38.],
        [193.,  39.],
        [181.,  40.],
        [183.,  41.],
        [218.,  42.],
        [230.,  43.],
        [242.,  44.],
        [209.,  45.],
        [1

### Memory management:

Managing memory well is crucial when working with neural architectures. Every operation can allocate new memory for its results. For example, B = A + B redirects tensor B to a new location containing the sum of tensor A + B, abandoning its old memory address. We can verify this simply by
using id(B), which returns an object’s address in memory. After an operation, id(B) will be in a new
location from before.

In [12]:
torch.manual_seed(42)
A = torch.randn(3,2)
B = torch.ones(3,2)
print(f'Original location in memory', id(B))
B = A + B
print(f'New location in memory', id(B))

Original location in memory 1645331832240
New location in memory 1645332506944


This can be problematic, as we want to avoid unnecessarily allocating memory, which is wasteful
given that we are often faced with large datasets and frequent updates in machine learning. Secondly,
parameters are often referenced in multiple places, so inconsistent updates across references can
cause issues.

In [13]:
A = torch.randn(3,2)
B = torch.ones(3,2)
print('Location in before operation', id(B))
B[:] = A + B
print('Location in after operation', id(B))

Location in before operation 1647117458064
Location in after operation 1647117458064



### A brief look at autograd

In PyTorch, every tensor operation adds to a tape, which is a record that Autograd uses to compute
gradients. The backward pass peels back this tape layer by layer, applying the JVPs in reverse sequence. This is the heart of backpropagation—the process which informs how we adjust parameters
to minimize loss. We can illustrate this with an example showing how autograd computes gradients
for a composition of logarithm and sine functions:

In [21]:
# Define operations of forward pass
def forward_pass(x):
    # Operation 1 - logarithm
    log_x = torch.log(x)
    # Operation 2 - sine
    sin_log_x = torch.sin(log_x)
    return sin_log_x

# Instantiate x tensor, setting requires_grad=True in order to track computations
x = torch.tensor([2.0], requires_grad=True)
# Forward pass to compute output
y = forward_pass(x)
# Backward pass, to compute gradient of y with respect to x
y.backward()
# Gradient stored in x.grad
print(f"Gradient of y with respect to x is: {x.grad}")

Gradient of y with respect to x is: tensor([0.3846])


Gradient of y with respect to x is tensor([0.3846])

In this code, autograd integrates the logarithmic and sine functions into the computational graph.
Upon invoking y.backward(), it employs the JVPs of these operations to calculate gradient of the
loss with respect to our input tensor x. This gradient is stored in x.grad, which is the culmination of
complex computations rendered accessible by autograd’s design.

Autograd’s utility is further exemplified in how it handles complex functions. Rather than constructing
voluminous Jacobian matrices, it computes Jacobian-vector products (JVPs)—a far more memory-efficient approach. This method showcases autograd’s computational efficiency and adaptability.

Whether dealing with a recurrent neural network or a feedforward model, autograd empowers us to
focus on model design and problem-solving. Manual gradient calculation has become unnecessary,
replaced by the dynamic computational pathways of autograd in PyTorch’s computational graphs.

### Practicing PyTorch fundamentals

In what remains of this section we are going to build some simple neural networks; we’ll discuss these
more in the next chapter. Here you are simply looking to add depth of understanding for the fundamentals we have already discussed. First, let’s expand on our previous gradient calculation example,
and apply it to what would be a single layer of a neural network. Here, we define some tensors: input,
weights, actual values to predict, and the output layer:

In [24]:
# Definitions - input data, weights, and 'true output' for loss calculation
input_tensor = torch.tensor([[1.0, 2.0, 3.0]], requires_grad=True)
weights = torch.tensor([[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]], requires_grad=True)
true_output = torch.tensor([[0.7, 1.5]])

# Forward pass - Matrix multiplication (input with weights), using ReLU activation
layer_output = torch.mm(input_tensor, weights.t()).relu()

# Calculate loss - MSE
predicted_output = layer_output
loss = (true_output - predicted_output).pow(2).mean()

# Backward pass with autograd to compute gradients
loss.backward()

# dict of tensors and gradients
nn_basic_results = {
 "Input Tensor": input_tensor,
 "Weights": weights,
 "Layer Output after ReLU": layer_output,
 "Loss (MSE)": loss,
 "Gradient with respect to Input Tensor": input_tensor.grad,
 "Gradient with respect to Weights": weights.grad
}

# print result
for key, value in nn_basic_results.items():
    print(f'{key}: {value}\n')

Input Tensor: tensor([[1., 2., 3.]], requires_grad=True)

Weights: tensor([[0.1000, 0.2000, 0.3000],
        [0.4000, 0.5000, 0.6000]], requires_grad=True)

Layer Output after ReLU: tensor([[1.4000, 3.2000]], grad_fn=<ReluBackward0>)

Loss (MSE): 1.6900001764297485

Gradient with respect to Input Tensor: tensor([[0.7500, 0.9900, 1.2300]])

Gradient with respect to Weights: tensor([[0.7000, 1.4000, 2.1000],
        [1.7000, 3.4000, 5.1000]])



In [25]:
from torch.utils.data import DataLoader, TensorDataset

# Define neural network
class SimpleNN(torch.nn.Module):
    def __init__(self):
        super(SimpleNN, self).__init__()
        self.weights = torch.nn.Parameter(torch.tensor([[0.1, 0.2, 0.3], [0.4, 0.5, 0.6]]))

    def forward(self, x):
        # Forward pass through layer and ReLU activation
        return torch.mm(x, self.weights.t()).relu()

# Instantiate neural network
model = SimpleNN()

# Define optimizer
optimizer = torch.optim.SGD(model.parameters(), lr=0.01)

# Define tensors for input and output
input_tensor = torch.tensor([[1.0, 2.0, 3.0]], requires_grad=True)
true_output = torch.tensor([[0.7, 1.5]])

# Wrap tensors in a DataLoader
train_loader = DataLoader(TensorDataset(input_tensor, true_output), batch_size=1)

# Training loop
for epoch in range(1):
    for batch in train_loader:
        # Separate data into inputs and outputs
        inputs, targets = batch

        # Zero gradients
        optimizer.zero_grad()

        # Forward pass
        predicted_output = model(inputs)

        # Compute loss
        loss = torch.mean((targets - predicted_output).pow(2))

        # Backward pass
        loss.backward()

        # Update weights
        optimizer.step()

        print("Loss: ", loss.item(), end="")

Loss:  1.6900001764297485

Notice that the loss is identical.

In [17]:
import lightning as L
from torch.utils.data import DataLoader, TensorDataset

class SimpleNN(L.LightningModule):
    def __init__(self):
        super(SimpleNN, self).__init__()
        self.weights = torch.nn.Parameter(torch.tensor([
        [0.1, 0.2, 0.3], [0.4, 0.5, 0.6]]))

    def forward(self, x):
        # Forward pass
        return torch.mm(x, self.weights.t()).relu()

    def training_step(self, batch, batch_idx):
        # Forward pass with loss calculation
        x, true_output = batch
        predicted_output = self(x)
        loss = torch.mean((true_output - predicted_output).pow(2))
        return loss

    def configure_optimizers(self):
        # Optimization with SGD
        return torch.optim.SGD(self.parameters(), lr=0.01)

# Instantiate neural network
model = SimpleNN()

# Define input and output tensors
input_tensor = torch.tensor([[1.0, 2.0, 3.0]])
true_output = torch.tensor([[0.7, 1.5]])

# Wrap data in a DataLoader
train_loader = DataLoader(TensorDataset(input_tensor, true_output),
 batch_size=1)

# Define trainer
trainer = L.Trainer(max_epochs=1)

# Fit model
trainer.fit(model, train_loader)

C:\Users\brend\Desktop\PythonVenv\py3_13\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
C:\Users\brend\Desktop\PythonVenv\py3_13\Lib\site-packages\lightning\pytorch\trainer\connectors\logger_connector\logger_connector.py:76: Starting from v1.9.0, `tensorboardX` has been removed as a dependency of the `lightning.pytorch` package, due to potential conflicts with other packages in the ML ecosystem. For this reason, `logger=True` will use `CSVLogger` as the default logger, unless the `tensorboard` or `tensorboardX` packages are found. Please `pip install lightning[extra]` or one of them to enable TensorBoard support by default
💡 Tip: For seamless cloud logging and experiment tracking, try installing [litlogger](https://pypi.org/project/li

┏━━━┳━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━━━━┳━━━━━━━┓
┃   ┃ Name         ┃ Type ┃ Params ┃ Mode ┃ FLOPs ┃
┡━━━╇━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━━━━╇━━━━━━━┩
│   │ other params │ n/a  │      6 │ n/a  │   n/a │
└───┴──────────────┴──────┴────────┴──────┴───────┘

Trainable params: 6                                                                                                
Non-trainable params: 0                                                                                            
Total params: 6                                                                                                    
Total estimated model params size (MB): 0.000                                                                      
Modules in train mode: 0                                                                                           
Modules in eval mode: 0                                                                                            
Total FLOPs: 0

C:\Users\brend\Desktop\PythonVenv\py3_13\Lib\site-packages\lightning\pytorch\utilities\_pytree.py:21: `isinstance(treespec, LeafSpec)` is deprecated, use `isinstance(treespec, TreeSpec) and treespec.is_leaf()` instead.
C:\Users\brend\Desktop\PythonVenv\py3_13\Lib\site-packages\lightning\pytorch\trainer\connectors\data_connector.py:434: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=15` in the `DataLoader` to improve performance.
C:\Users\brend\Desktop\PythonVenv\py3_13\Lib\site-packages\lightning\pytorch\loops\fit_loop.py:321: The number of training batches (1) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.


C:\Users\brend\Desktop\PythonVenv\py3_13\Lib\site-packages\rich\live.py:260: UserWarning: install "ipywidgets" for 
Jupyter support
  warnings.warn('install "ipywidgets" for Jupyter support')

`Trainer.fit` stopped: `max_epochs=1` reached.
